add hidden cascade delete location endpoint parameter

adamkorynta opened the pull request:
Fixed #342

adamkorynta committed:
add hidden cascade delete location endpoint parameter

adamkorynta commented:
@MikeNeilson - I'm not sure if you were suggesting that a role check be added here at the CDA level or if that should get added to the CWMS schema. I'm also not sure if that would be a new role or using the PD account.

MikeNeilson commented:
I'm not really sure where the best place is, but I think this approach is suitable for now. It at least provides an extra barrier to entry to anyone who would currently be authorized to perform this operation.

adamkorynta commented:
Ok, I'll proceed with adding some tests

adamkorynta committed:
add tests for location cascade delete

this adds a new error handler for when a location delete fails due to foreign keyed data

adamkorynta commented:
For a failing location delete (non-cascade) I added a 405 error code. It seemed that or 409 seemed applicable, but I'm definitely interested in thoughts.

adamkorynta commented on cwms_radar_api/src/main/java/cwms/radar/api/LocationController.java:
```diff
@@ -352,21 +369,19 @@ public void update(Context ctx, @NotNull String locationId) {
-            ctx.status(HttpServletResponse.SC_INTERNAL_SERVER_ERROR).json(re);
+        } catch (DataAccessException ex) {
+            SQLException cause = ex.getCause(SQLException.class);
+            if(cause != null && cause.getErrorCode() == 20031) {
```
error code 20031 is specific to deleting locations so I didn't put this up in the general ApiServlet handling

MikeNeilson replied:
Makes sense. though it might be worth organizing all of those db error codes into a specific class at some point.

adamkorynta replied:
a mechanism for codgen would be nice

MikeNeilson replied:
Honestly wouldn't be too hard. or really even need codegen (if you meant JOOQ), the errors are already listed as rows in a table.

adamkorynta replied:
Yeah I was thinking of generating enums for some of the CWMS owned static tables: https://stackoverflow.com/questions/46429671/generate-enum-class-from-table-with-jooq

MikeNeilson replied:
Alternatively that code could get moved to the wrapException handler. But in this case this does seem pretty specific to this usage.

MikeNeilson commented:
> For a failing location delete (non-cascade) I added a 405 error code. It seemed that or 409 seemed applicable, but I'm definitely interested in thoughts.

Looking at the 405 docs an "allow" header is required in response so I don't think it's quite right. Method in that context does generally mean the HTTP VERB, not our API operation.

409 Conflict makes more sense since desired operation is in conflict with the state of things.

adamkorynta commented:
Works for me, I kept seeing 409 as being for POST

adamkorynta committed:
update to use a 409 conflict error code when unable to delete

MikeNeilson commented on cwms_radar_api/src/test/java/cwms/radar/api/LocationControllerTestIT.java:
```diff
@@ -0,0 +1,147 @@
+
+        // delete without cascade should fail
+        given()
+                .config(RestAssured.config().jsonConfig(jsonConfig().numberReturnType(JsonPathConfig.NumberReturnType.DOUBLE)))
```
This should be the default, was it giving you trouble without it?

adamkorynta replied:
yeah worked without it, I just copy pasted from elsewhere

MikeNeilson commented on cwms_radar_api/src/test/java/cwms/radar/api/LocationControllerTestIT.java:
```diff
@@ -0,0 +1,147 @@
+                .assertThat()
+                .statusCode(is(HttpServletResponse.SC_CONFLICT));
+
+        // delete without cascade should fail
```
Incorrect comment.

adamkorynta replied:
fixed

MikeNeilson reviewed (changes requested):
A few nitpicks and a question.

adamkorynta committed:
clean up test cases

adamkorynta committed:
re-enable test loading scripts
